### DISCUSSION TITLE
Tests with incremental parsing

### DISCUSSION THREAD

**Discussion by uben0** [tree-sitter/tree-sitter#2942] [Q&A]
Is it possible, in a test, to simulate incremental parsing with character insertion?
```
=====
my awesome test
=====
3 + 1
=====[0, 5]
* 2
-----
(add (int) (int))
(add (int) (mul (int) (int))
```

I develop and maintain a *Tree-sitter* grammar https://github.com/uben0/tree-sitter-typst and the issue https://github.com/uben0/tree-sitter-typst/issues/23 only appears in incremental parsing. It's very hard to understand what's going on without the this testing feature.

If this feature doesn't exist, is there a work around or a technique? More in general what are the debug tool provided by *Tree-sitter*?

At the moment, I use `tree-sitter test` with `-d` or `-D`, and `tree-sitter parse` to see nodes' ranges. By the way, is there a way to specify nodes' ranges in the expected tree?
```
=====
my awesome test
=====
foo
-----
(ident [0, 0] - [0, 3])
```

**Comment by amaanq** [comment]
There's no easy way to simulate incremental parsing via the test format, but you can easily do so yourself if you're familiar with nodejs or rust in your grammar repo. You can also "fuzz" edits, take a look here to see how we test that for every upstream language: <https://github.com/tree-sitter/tree-sitter/blob/7147c727371a3245a2580e407a0b1e541afbb972/cli/src/tests/corpus_test.rs#L210>
